sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -1708,6 +2044,275 @@ a normal SYCL command-group submission.
+a normal SYCL command-group submission.
+
+* Throws synchronously with error code `invalid` if the `sycl::event` associated
+with `depEvent` came from the same graph that the graph node resulting from this
+command-group submission is associated with.
+
+* Throws synchronously with error code `invalid` if `depEvent` is a default
+constructed `dynamic_event` with no underlying SYCL event.
+
+|
+[source,c++]
+----
+void handler::depends_on(const std::vector<dynamic_event>& depEvents);
+----
+
+|Overload of `depends_on()` which takes a list of dynamic event dependencies.
+
+This function has the same semantics as `depends_on(const std::vector<event>&)`.
+
+Parameters:
+
+* `depEvents` - List of dynamic event dependencies.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if this function is called from
+a normal SYCL command-group submission.
+
+* Throws synchronously with error code `invalid` if any of the `sycl::event`
+objects associated with `depEvents` came from the same graph that the graph node
+resulting from this command-group submission is associated with.
+
+* Throws synchronously with error code `invalid` if any of `depEvents` is a
+default constructed `dynamic_event` with no underlying SYCL event.
+|===
+
+=== Events
+
+Events can be used with graphs in the following ways:
+
+- Defining dependencies between nodes in the same graph.
+- Defining dependencies between <<inter-graph-dependencies, different graphs>>.
+- Obtaining <<node-execution-events, execution events for invidual nodes>>
+within an executable graph, which can be waited on or used as dependencies for
+eager SYCL operations.
+- Creating external event dependencies for a graph, either
+<<external-event-dependencies, one-time>> or <<dynamic-events, updatable between
+executions>>.
+
+==== Node-Level Execution Events [[node-execution-events]]
+
+Events representing the completion of an individual node inside a given
+executable graph can be obtained using
+`command_graph<graph_state::executable>::get_event(node)`. These events can then
+be waited on or used as dependencies for eager SYCL operations outside of
+graphs. These events may be useful for operations which may be infrequent and
+depend only on some intermediate results of work being done in the graph.
+
+[source, c++]
+....
+sycl::event ExecutionEvent = ExecGraph.get_event(SomeNode);
+
+Queue.submit((sycl::handler& CGH)
+  {
+    CGH.depends_on(ExecutionEvent);
+    CGH.parallel_for(...);
+  });
+
+// The above operation will only execute once SomeNode has finished executing
+// inside ExecGraph
+Queue.ext_oneapi_graph(ExecGraph);
+....
+
+These events represent only the most recent execution of a given executable
+graph. If an application executes the same graph multiple times before
+scheduling work or performing a host-side wait on the event then executions of
+the node in a previous execution other than the most recent one may be missed.
```

Comment:
for the long sentence, does this talk about the code like below?
```
sycl::event ExecutionEvent = ExecGraph.get_event(SomeNode);

for (int i = 0; i < 1000; ++i)  // 1000 is just one value for example
    Queue.ext_oneapi_graph(ExecGraph);

Queue.submit((sycl::handler& CGH)
{
    CGH.depends_on(ExecutionEvent);
    CGH.parallel_for(...);
});
```

what is "Queue.ext_oneapi_graph(ExecGraph); with i==998" missed?

Reply:
After having some internal discussions with regards to implementability of this aspect of the spec I've pushed some wording to clarify what these events represent, but I'll also address your specific example.

As it stands right now if multiple SYCL graph submissions are in-flight as in your example, it is undefined at the point where you call `CGH.depends_on(ExecutionEvent);` which exact execution it will depend on, as it depends which iteration is currently executing on device.

To depend on a specific execution the application must enforce ordering such that multiple enqueues are not in-flight, like so (example I just added taken from the spec):

```c++
sycl::event NodeExecutionEvent;
sycl::event GraphCompletionEvent;

for (size_t i = 0; i < GraphIterations; i++){
  // Obtain the node execution event for the graph
  NodeExecutionEvent = ExecGraph.get_event(SomeNode);

  // Enqueue the graph for execution
  sycl::event GraphCompletionEvent = Queue.ext_oneapi_graph(ExecGraph);
  // Use the event, to wait on the host and perform some intermediate
  // host-work once that node has completed
  ExecutionEvent.wait();
  DoSomethingOnHost();
  // Wait on the graph finishing to ensure it is complete before the next use of
  // a node-level execution event
  GraphCompletionEvent.wait_and_throw();
}
```

This is generally in keeping with how the backends handle this but slightly more restrictive because SYCL events do not function like CUDA/L0 events and the user does not have actual visibility of device execution because this is obfuscated by the SYCL runtime.